engine/lib/phx/src/engine/event_bus.rs

```diff
@@ -0,0 +1,303 @@
+    pub subscribers: Vec<Subscriber>,
+    // pub payloads: Vec<EventPayload>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct MessageRequest {
+    priority: u16,
+    name: String,
+    stay_alive: bool,
+    completed: bool,
+}
+
+impl Ord for MessageRequest {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        other
+            .priority
+            .cmp(&self.priority)
+            .then_with(|| self.name.cmp(&other.name))
+    }
+}
+
+impl PartialOrd for MessageRequest {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub struct EventBus {
+    events: HashMap<String, Event>,
+    update_pass_map: HashMap<UpdatePass, BinaryHeap<MessageRequest>>,
+    next_tunnel_id: AtomicU32,
+    max_priority_locked: bool,
+}
+
+impl EventBus {
+    pub fn new() -> Self {
+        Self {
+            next_tunnel_id: AtomicU32::new(0),
+            update_pass_map: HashMap::new(),
+            events: HashMap::new(),
+            max_priority_locked: false,
+        }
+    }
+
+    pub fn lock_max_priority(&mut self) {
+        self.max_priority_locked = true;
+    }
+
+    pub fn dispatch(&mut self, update_pass: UpdatePass, lua: &mut Ref<Lua>) {
+        //* Can we handle the lua state differently? */
+        //* Maybe also with future potential multiple lua states in mind */
+        let globals = lua.globals();
+        let event_tunnels: Table = globals.get("EventTunnels").expect("Unknown table");
+
+        if let Some(message_heap) = self.update_pass_map.get_mut(&update_pass) {
+            let mut message_requests: Vec<_> = message_heap.drain().collect();
+            message_requests.sort_by(|a, b| b.priority.cmp(&a.priority));
+
+            for message_request in &mut message_requests {
+                if let Some(event) = self.events.get(&message_request.name) {
+                    for subscriber in &event.subscribers {
+                        let id = subscriber.tunnel_id;
+
+                        if !message_request.completed && !message_request.stay_alive {
+                            println!(
+                                "Dispatched event {} through tunnel with id {}",
+                                message_request.name, id
+                            )
+                        }
+
+                        // todo: only dispatch to subscriber with set entity_id if defined
+                        let tunnel_func: Function = event_tunnels
+                            .get(id)
+                            .expect(&format!("Unknown tunnel with id: {}", id));
+
+                        if let Err(e) = tunnel_func.call::<_, ()>(()) {
+                            trace!("{}", e);
+                        } else {
+                            message_request.completed = true; // Mark as completed
+                        }
+                    }
+                } else {
+                    panic!("Event not found: {}", message_request.name);
```

Comment:
Is a panic really necessary? Should we just warn and continue?

Reply:
Imo engine events shouldn`t be able to fail. Else the developer has made some error. That is why i have a panic here and at other places. There should not be any floating around events.

Reply:
Understood, thanks